### ISSUE TITLE
Autograd Warning

### ISSUE THREAD

**Issue by akshay-gupta123** [frgfm/torch-cam#72]
## 🐛 Bug
When I am integrating densenet121 with torchcam I am getting following warning, which I am not getting while using the model diectly. 

` usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1204: UserWarning: Output 0 of BackwardHookFunctionBackward is a view and is being modified inplace. This view was created inside a custom Function (or because an input was returned as-is) and the autograd logic to handle view+inplace would override the custom backward associated with the custom Function, leading to incorrect gradients. This behavior is deprecated and will be forbidden starting version 1.6. You can remove this warning by cloning the output of the custom Function. (Triggered internally at  /pytorch/torch/csrc/autograd/variable.cpp:547.)
  result = torch.relu_(input)
`

**Here is the code**

```
model = densenet121(pretrained=True).eval()

img = read_image("image.png")
input_tensor = normalize(resize(img, (224, 224)) / 255., [0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
extract = GradCAM(model=model)
out = model(input_tensor.unsqueeze(0))
```
## Environment
- I am working on Google colab
- Installed Torchcam using pip
- Pytorch version: 1.8.1+cu101 
- Torchcam Version: 0.2.0
- Python version: 3.7.10

**Comment by frgfm** [comment]
Hello @akshay-gupta123 :wave: 

Thanks for reporting this! This is indeed a strange behaviour, which seems linked to the core autograd mechanism of PyTorch. I tried a few things quickly this morning to solve this but didn't succeed.

I'll investigate, but this is most likely due to some specific ops conducted in the densenet architecture (this issue isn't happening on resnet)

**Comment by frgfm** [comment]
Quick update: I made a few tests on my own, I believe this is due to the inplace ReLU in densenet over here https://github.com/pytorch/vision/blob/master/torchvision/models/densenet.py#L217

**Comment by akshay-gupta123** [comment]
Ya it's working fine for resnet, but same warning occurs with squeezenet1_0. May be it's due to some architecture ops in torchvision models.

**Comment by frgfm** [comment]
Hi @akshay-gupta123,

I didn't forget the problem, and now it's not only throwing a warning but an error in recent pytorch versions. I opened an issue on their repo trying to make sense of this which seems to be an issue with backward hooks!

**Comment by akshay-gupta123** [comment]
Hi @frgfm 
Thanks for update. I made a forum in this regard about a month ago but sadly doesn't gets anybody's attention.

**Comment by wizlee** [comment]
Hi @frgfm ,
Is there any workaround while you are working on the fix? Maybe something like another branch with the fix where I can install from source?

**Comment by frgfm** [comment]
Hi @wizlee :wave: 

If you're talking about in-place activation modules in general, yes there is: switching their `inplace` attribute to False.
So this would work
```python
import torch
from torch import nn
# Get the model
mod = nn.Sequential(nn.Conv2d(3, 8, 3, padding=1), nn.ReLU(), nn.Conv2d(8, 8, 3, padding=1), nn.ReLU(inplace=True))
# Hook a module before the in place ReLU
def hook_g(m, input, output): print(input.shape)
mod[2].register_full_backward_hook(hook_g)
# Inference
mod[-1].inplace = False
out = mod(torch.rand((1, 3, 224, 224)))
```

But more specifically for densenet121, the issue is that this isn't a module but the functional form. So either override the forward or convert it to a sequential
 ```python
import torch
from torch import nn
from torchvision.models import densenet121
from torchvision.io import read_image
from torchvision.transforms.functional import resize, normalize
from torchcam.cams import GradCAM
# Get the model
mod = densenet121(pretrained=True).eval()
# Recreating it as a sequential (notice the ReLU isn't in place this time)
mod = nn.Sequential(mod.features, nn.ReLU(), nn.AdaptiveAvgPool2d((1, 1)), nn.Flatten(1), mod.classifier)

img = read_image("image.png")
input_tensor = normalize(resize(img, (224, 224)) / 255., [0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
extract = GradCAM(model=mod)
out = mod(input_tensor.unsqueeze(0))
```

Let me know if that helps :+1:

**Comment by frgfm** [comment]
Actually, I ended up digging a bit some tricks that were suggested to me and got it working. The PR is now merged in the master branch @wizlee @akshay-gupta123 :raised_hands: 

Feel free to reopen if you still encounter this issue!

**Comment by wizlee** [comment]
Thanks @frgfm! Will try out the latest. Glad that you perform the fix promptly as the only modification to model I did is to replace the output layer for fine-tuning. So there's a high chance that I will face stumbling blocks when trying your previous workaround 😉

**Comment by frgfm** [comment]
The release is out :balloon:
Bunch of new features, everything is detailed in the release note: https://github.com/frgfm/torch-cam/releases/tag/v0.3.0
